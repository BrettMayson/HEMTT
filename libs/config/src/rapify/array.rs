use std::iter::Sum;

use byteorder::ReadBytesExt;
use hemtt_common::io::{ReadExt, WriteExt, compressed_int_len};

use crate::{Array, Item, Number, Str};

use super::{Derapify, Rapify};

impl Rapify for Array {
    fn rapify<O: std::io::Write>(
        &self,
        output: &mut O,
        offset: usize,
    ) -> Result<usize, std::io::Error> {
        let mut written = output.write_compressed_int(self.items.len() as u32)?;
        for item in &self.items {
            output.write_all(&[item.rapified_code()])?;
            written += item.rapify(output, offset)? + 1;
        }
        assert_eq!(written, self.rapified_length());
        Ok(written)
    }

    fn rapified_length(&self) -> usize {
        compressed_int_len(self.items.len() as u32)
            + usize::sum(self.items.iter().map(|e| e.rapified_length() + 1))
    }
}

impl Array {
    /// Derapify an array from the input stream
    ///
    /// # Errors
    /// [`std::io::Error`] if the input stream is invalid or cannot be read
    pub fn derapify<I: std::io::Read + std::io::Seek>(
        input: &mut I,
        expand: bool,
    ) -> Result<Self, std::io::Error> {
        let start = input.stream_position()? as usize;
        let length = input.read_compressed_int()?;
        let mut items = Vec::with_capacity(length as usize);
        for _ in 0..length {
            let item = Item::derapify(input)?;
            items.push(item);
        }
        Ok(Self {
            items,
            span: start..input.stream_position()? as usize,
            expand,
        })
    }
}

impl Rapify for Item {
    fn rapify<O: std::io::Write>(
        &self,
        output: &mut O,
        offset: usize,
    ) -> Result<usize, std::io::Error> {
        match self {
            Self::Str(s) => s.rapify(output, offset),
            Self::Number(n) => n.rapify(output, offset),
            Self::Array(a) => {
                let mut written = output.write_compressed_int(a.len() as u32)?;
                for item in a {
                    output.write_all(&[item.rapified_code()])?;
                    written += item.rapify(output, offset)? + 1;
                }
                Ok(written)
            }
            Self::Invalid(_) => unreachable!(),
        }
    }

    fn rapified_length(&self) -> usize {
        match self {
            Self::Str(s) => s.rapified_length(),
            Self::Number(n) => n.rapified_length(),
            Self::Array(a) => {
                compressed_int_len(a.len() as u32)
                    + usize::sum(a.iter().map(|e| e.rapified_length() + 1))
            }
            Self::Invalid(_) => unreachable!(),
        }
    }

    fn rapified_code(&self) -> u8 {
        match self {
            Self::Str(s) => s.rapified_code(),
            Self::Number(n) => n.rapified_code(),
            Self::Array(_) => 3,
            Self::Invalid(_) => unreachable!(),
        }
    }
}

impl Derapify for Item {
    fn derapify<I: std::io::Read + std::io::Seek>(input: &mut I) -> Result<Self, std::io::Error>
    where
        Self: Sized,
    {
        let code = input.read_u8()?;
        match code {
            0 => Ok(Self::Str(Str::derapify(input)?)),
            1 => Ok(Self::Number(Number::derapify_float32(input)?)),
            2 => Ok(Self::Number(Number::derapify_int32(input)?)),
            3 => {
                let length = input.read_compressed_int()?;
                let mut items = Vec::with_capacity(length as usize);
                for _ in 0..length {
                    let item = Self::derapify(input)?;
                    items.push(item);
                }
                Ok(Self::Array(items))
            }
            6 => Ok(Self::Number(Number::derapify_int64(input)?)),
            _ => Err(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                format!("Invalid rapified item code: {code}"),
            )),
        }
    }
}
