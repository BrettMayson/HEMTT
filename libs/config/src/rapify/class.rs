use std::io::Cursor;

use byteorder::{LittleEndian, WriteBytesExt};
use hemtt_common::io::{compressed_int_len, WriteExt};

use crate::{Class, Ident, Property};

use super::Rapify;

impl Rapify for Class {
    fn rapify<O: std::io::Write>(
        &self,
        output: &mut O,
        offset: usize,
    ) -> Result<usize, std::io::Error> {
        let mut written = 0;
        match self {
            Self::Local { properties, .. } | Self::Root { properties } => {
                let parent = self.parent();
                if let Some(parent) = &parent {
                    output.write_cstring(parent.as_str())?;
                    written += parent.as_str().len() + 1;
                } else {
                    written += output.write(b"\0")?;
                }
                written += output.write_compressed_int(properties.len() as u32)?;

                let properties_len = properties
                    .iter()
                    .map(|p| p.name().len() + 1 + p.rapified_length())
                    .sum::<usize>();
                let mut class_offset = offset + written + properties_len + 4;
                let mut class_bodies: Vec<Cursor<Box<[u8]>>> = Vec::new();
                let pre_properties = written;

                for property in properties {
                    let pre_write = written;
                    let code = property.property_code();
                    output.write_all(&code)?;
                    written += code.len();
                    output.write_cstring(property.name().as_str())?;
                    written += property.name().len() + 1;
                    match property {
                        Property::Entry { value, .. } => {
                            written += value.rapify(output, offset)?;
                        }
                        Property::Class(c) => {
                            if let Self::Local { .. } = c {
                                output.write_u32::<LittleEndian>(class_offset as u32)?;
                                written += 4;
                                let buffer: Box<[u8]> =
                                    vec![0; c.rapified_length()].into_boxed_slice();
                                let mut cursor = Cursor::new(buffer);
                                let body_size = c.rapify(&mut cursor, class_offset)?;
                                assert_eq!(body_size, c.rapified_length());
                                class_offset += body_size;
                                class_bodies.push(cursor);
                            }
                        }
                        Property::Delete(_) => continue,
                        Property::MissingSemicolon(_, _) => unreachable!(),
                    }
                    assert_eq!(
                        written - pre_write,
                        property.rapified_length() + property.name().len() + 1
                    );
                }

                assert_eq!(written - pre_properties, properties_len);

                output.write_u32::<LittleEndian>(class_offset as u32)?;
                written += 4;

                for cursor in class_bodies {
                    output.write_all(cursor.get_ref())?;
                    written += cursor.get_ref().len();
                }
            }
            Self::External { name } => {
                output.write_all(&[3])?;
                output.write_cstring(name.as_str())?;
                written += 1;
            }
        }
        Ok(written)
    }

    fn rapified_length(&self) -> usize {
        match self {
            Self::External { .. } => 0,
            Self::Local { properties, .. } | Self::Root { properties, .. } => {
                let parent_length = self.parent().map_or(0, Ident::len);
                parent_length
                    + 1 // parent null terminator
                    + 4 // offset to next class
                    + compressed_int_len(properties.len() as u32)
                    + properties
                        .iter()
                        .map(|p| {
                            p.name().len()
                                + 1 // name null terminator
                                + p.rapified_length()
                                + match p {
                                    Property::Class(c) => c.rapified_length(),
                                    _ => 0,
                                }
                        })
                        .sum::<usize>()
            }
        }
    }
}
