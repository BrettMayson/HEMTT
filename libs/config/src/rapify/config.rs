use std::io::Cursor;

use byteorder::{LittleEndian, WriteBytesExt};

use crate::{Class, Config};

use super::Rapify;

impl Rapify for Config {
    fn rapify<O: std::io::Write>(
        &self,
        output: &mut O,
        _offset: usize,
    ) -> Result<usize, std::io::Error> {
        output.write_all(b"\0raP")?;
        output.write_all(b"\0\0\0\0\x08\0\0\0")?;

        let root_class = Class::Root {
            properties: self.0.clone(),
        };
        let buffer: Box<[u8]> = vec![0; root_class.rapified_length()].into_boxed_slice();
        let mut cursor = Cursor::new(buffer);
        let written = root_class.rapify(&mut cursor, 16)?;
        assert_eq!(written, root_class.rapified_length());

        let enum_offset = 16 + cursor.get_ref().len() as u32;
        output.write_u32::<LittleEndian>(enum_offset)?;

        output.write_all(cursor.get_ref())?;

        output.write_all(b"\0\0\0\0")?;
        assert_eq!(written + 20, self.rapified_length());
        Ok(written + 20)
    }

    fn rapified_length(&self) -> usize {
        let root_class = Class::Root {
            properties: self.0.clone(),
        };
        root_class.rapified_length() + 20 // metadata
    }
}
