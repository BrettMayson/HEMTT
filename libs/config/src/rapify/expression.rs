use hemtt_common::io::{ReadExt, WriteExt};

use crate::Expression;

use super::{Derapify, Rapify};

impl Rapify for Expression {
    fn rapify<O: std::io::Write>(
        &self,
        output: &mut O,
        _offset: usize,
    ) -> Result<usize, std::io::Error> {
        output.write_cstring(&self.value)?;
        Ok(self.value.len() + 1)
    }

    fn rapified_length(&self) -> usize {
        self.value.len() + 1
    }

    fn rapified_code(&self) -> u8 {
        4
    }
}

impl Derapify for Expression {
    fn derapify<I: std::io::Read + std::io::Seek>(input: &mut I) -> Result<Self, std::io::Error>
    where
        Self: Sized,
    {
        let start = input.stream_position()? as usize;
        let value = input.read_cstring()?;
        Ok(Self {
            value,
            span: start..input.stream_position()? as usize,
        })
    }
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
mod tests {
    use crate::Expression;

    use super::Rapify;

    #[test]
    fn str() {
        let mut buffer = Vec::new();
        let written = Expression {
            value: "getResolution".to_string(),
            span: 0..14,
        }
        .rapify(&mut buffer, 0)
        .unwrap();
        assert_eq!(written, 14);
        assert_eq!(
            buffer,
            vec![
                103, 101, 116, 82, 101, 115, 111, 108, 117, 116, 105, 111, 110, 0
            ]
        );
    }
}
