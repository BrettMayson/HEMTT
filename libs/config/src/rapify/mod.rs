//! Rapify configs into a binary format

mod array;
mod class;
mod config;
mod expression;
mod number;
mod property;
mod str;
mod value;

use std::io::Write;

/// Trait for rapifying objects
pub trait Rapify {
    /// Rapify the object into the output stream
    ///
    /// # Errors
    /// if the output stream fails
    fn rapify<O: Write>(&self, output: &mut O, offset: usize) -> Result<usize, std::io::Error>;
    /// Get the length of the rapified object
    fn rapified_length(&self) -> usize;
    /// Get the rapified element code
    fn rapified_code(&self) -> u8 {
        3
    }
}
