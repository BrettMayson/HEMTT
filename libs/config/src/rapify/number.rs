use byteorder::{LittleEndian, WriteBytesExt};

use crate::Number;

use super::Rapify;

impl Rapify for Number {
    fn rapify<O: std::io::Write>(
        &self,
        output: &mut O,
        _offset: usize,
    ) -> Result<usize, std::io::Error> {
        let mut written = 0;
        match self {
            Self::Int32 { value, .. } => {
                output.write_i32::<LittleEndian>(*value)?;
                written += 4;
            }
            Self::Int64 { value, .. } => {
                output.write_i64::<LittleEndian>(*value)?;
                written += 8;
            }
            Self::Float32 { value, .. } => {
                output.write_f32::<LittleEndian>(*value)?;
                written += 4;
            }
        }
        Ok(written)
    }

    fn rapified_length(&self) -> usize {
        match self {
            Self::Int32 { .. } | Self::Float32 { .. } => 4,
            Self::Int64 { .. } => 8,
        }
    }

    fn rapified_code(&self) -> u8 {
        match self {
            Self::Int32 { .. } => 2,
            Self::Int64 { .. } => 6,
            Self::Float32 { .. } => 1,
        }
    }
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
mod tests {
    use crate::Number;

    use super::Rapify;

    #[test]
    fn int32() {
        let mut buffer = Vec::new();
        let written = Number::Int32 {
            value: 1234,
            span: 0..4,
        }
        .rapify(&mut buffer, 0)
        .unwrap();
        assert_eq!(written, 4);
        assert_eq!(buffer, vec![0xd2, 0x04, 0x00, 0x00]);
    }

    #[test]
    fn int64() {
        let mut buffer = Vec::new();
        let written = Number::Int64 {
            value: 1234,
            span: 0..4,
        }
        .rapify(&mut buffer, 0)
        .unwrap();
        assert_eq!(written, 8);
        assert_eq!(buffer, vec![0xd2, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]);
    }

    #[test]
    fn float32() {
        let mut buffer = Vec::new();
        let written = Number::Float32 {
            value: 1234.0,
            span: 0..4,
        }
        .rapify(&mut buffer, 0)
        .unwrap();
        assert_eq!(written, 4);
        assert_eq!(buffer, vec![0x00, 0x40, 0x9A, 0x44]);
    }
}
