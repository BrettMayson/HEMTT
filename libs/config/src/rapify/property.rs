use crate::{Class, Property, Value};

use super::Rapify;

impl Rapify for Property {
    fn rapify<O: std::io::Write>(
        &self,
        _output: &mut O,
        _offset: usize,
    ) -> Result<usize, std::io::Error> {
        unreachable!()
    }

    fn rapified_length(&self) -> usize {
        self.property_code().len()
            + match self {
                Self::Entry { value, .. } => match value {
                    Value::Str(s) => s.rapified_length(),
                    Value::Number(n) => n.rapified_length(),
                    Value::Array(a) => a.rapified_length(),
                    Value::UnexpectedArray(_) | Value::Invalid(_) => unreachable!(),
                },
                Self::Class(c) => match c {
                    Class::Local { .. } => 4,
                    Class::External { .. } => 0,
                },
                Self::Delete(_) => 0,
                Self::MissingSemicolon(_, _) => unreachable!(),
            }
    }
}

impl Property {
    /// Get the code of the property
    #[must_use]
    pub fn property_code(&self) -> Vec<u8> {
        match self {
            Self::Entry { value, .. } => match value {
                Value::Str(s) => vec![1, s.rapified_code()],
                Value::Number(n) => vec![1, n.rapified_code()],
                Value::Array(a) => {
                    if a.expand {
                        vec![5, 1, 0, 0, 0]
                    } else {
                        vec![2]
                    }
                }
                Value::UnexpectedArray(_) | Value::Invalid(_) => unreachable!(),
            },
            Self::Class(c) => match c {
                Class::Local { .. } => vec![0],
                Class::External { .. } => vec![3],
            },
            Self::Delete(_) => {
                vec![4]
            }
            Self::MissingSemicolon(_, _) => unreachable!(),
        }
    }
}
