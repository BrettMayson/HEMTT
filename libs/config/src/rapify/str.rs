use hemtt_io::WriteExt;

use crate::Str;

use super::Rapify;

impl Rapify for Str {
    fn rapify<O: std::io::Write>(
        &self,
        output: &mut O,
        _offset: usize,
    ) -> Result<usize, std::io::Error> {
        output.write_cstring(&self.value)?;
        Ok(self.value.len() + 1)
    }

    fn rapified_length(&self) -> usize {
        self.value.len() + 1
    }

    fn rapified_code(&self) -> u8 {
        0
    }
}

#[cfg(test)]
mod tests {
    use crate::Str;

    use super::Rapify;

    #[test]
    fn str() {
        let mut buffer = Vec::new();
        let written = Str {
            value: "Hello World".to_string(),
            span: 0..12,
        }
        .rapify(&mut buffer, 0)
        .unwrap();
        assert_eq!(written, 12);
        assert_eq!(
            buffer,
            vec![0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x20, 0x57, 0x6f, 0x72, 0x6c, 0x64, 0x00]
        );
    }
}
