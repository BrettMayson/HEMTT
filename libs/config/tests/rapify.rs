#![allow(clippy::unwrap_used)]

use std::io::Read;

use hemtt_config::rapify::Rapify;
use hemtt_preprocessor::Processor;
use hemtt_workspace::LayerType;

const ROOT: &str = "tests/rapify/";

macro_rules! bootstrap {
    ($dir:ident) => {
        paste::paste! {
            #[test]
            fn [<config_rapify_ $dir>]() {
                check(stringify!($dir));
            }
        }
    };
}

fn check(dir: &str) {
    let folder = std::path::PathBuf::from(ROOT).join(dir);
    let workspace = hemtt_workspace::Workspace::builder()
        .physical(&folder, LayerType::Source)
        .finish(None, false, &hemtt_common::config::PDriveOption::Disallow)
        .unwrap();
    let source = workspace.join("source.hpp").unwrap();
    let processed = Processor::run(&source).unwrap();
    let parsed = hemtt_config::parse(None, &processed);
    let workspacefiles = hemtt_workspace::reporting::WorkspaceFiles::new();
    if let Err(e) = &parsed {
        let e = e
            .iter()
            .map(|e| e.diagnostic().unwrap().to_string(&workspacefiles))
            .collect::<Vec<_>>();
        std::fs::write(folder.join("stderr.ansi"), e.join("\n")).unwrap();
        std::fs::write(folder.join("processed.txt"), processed.as_str()).unwrap();
        panic!("failed to parse")
    };
    let parsed = parsed.unwrap();
    let mut expected = Vec::new();
    let expected_path = folder.join("expected.bin");
    if !expected_path.exists() {
        let mut file = std::fs::File::create(&expected_path).unwrap();
        parsed.config().rapify(&mut file, 0).unwrap();
        panic!("expected file did not exist, created it");
    };
    std::fs::File::open(expected_path)
        .unwrap()
        .read_to_end(&mut expected)
        .unwrap();
    let mut output = Vec::new();
    let written = parsed.config().rapify(&mut output, 0).unwrap();
    assert_eq!(written, parsed.config().rapified_length());
    assert_eq!(output, expected);
    let vanilla_path = folder.join("cfgconvert.bin");
    if vanilla_path.exists() {
        let mut expected = Vec::new();
        let mut file = std::fs::File::open(&vanilla_path).unwrap();
        file.read_to_end(&mut expected).unwrap();
        assert_eq!(output, expected);
    };
}

bootstrap!(ace_main);
bootstrap!(cba_multiline);
bootstrap!(delete_class);
bootstrap!(eval);
bootstrap!(external_class);
bootstrap!(inheritence_array_extend);
bootstrap!(join_digit);
bootstrap!(join_in_ident);
bootstrap!(join);
bootstrap!(nested_array);
bootstrap!(numbers);
bootstrap!(procedural_texture);
bootstrap!(single_class);
