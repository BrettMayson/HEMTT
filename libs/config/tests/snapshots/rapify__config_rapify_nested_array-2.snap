---
source: libs/config/tests/rapify.rs
expression: config.to_string()
---
value[] = {{1}, {2}};
value2[] = {{"test"}, "hello", {"hello", "hello"}};
modes[] = {
    "Single",
    "manual",
    "close",
    "short",
    "medium",
    "far_optic1",
    "far_optic2",
    "ACE_Burst_far"
};
values[] += {1};
